src/R/lib_size_sc_V5_single_species.R

```diff
@@ -75,36 +79,92 @@ duprate <- mean(Df[Df[, ncol+2] > 500,ncol+1])
+                 UNMAPPED_READS=0,
+                 UNPAIRED_READ_DUPLICATES=0,
+                 READ_PAIR_DUPLICATES= dupreads,
+                 READ_PAIR_OPTICAL_DUPLICATES=0,
+                 PERCENT_DUPLICATION=duprate,
+                 ESTIMATED_LIBRARY_SIZE=alllibeize)
 fileout <- paste(basename(filt), 'dups.log.txt', sep=".")
 write.table(out, fileout, quote=F, row.names=F, col.names=T, sep="\t")
 
 # Plot lib size
+options(scipen=999)
 Size2 <-Df[, ncol+2];  Size2 <- Size2[Size2 >0]; # plot(Size)
 file6 <- paste(basename(filt), 'libSize.png', sep=".")
 png(file6, width = 8, height = 8, units = 'in', res = 300)
 par(mfrow = c(2,1))
 Size2 <- sort(Size2, decreasing = T)
-elbow <- elbow_finder(1:length(Size2), log10(Size2))
-plot(Size2, log="y", xlab=paste0(" Barcode rank (", length(Size2)-elbow[1], " low quality cells)"), ylab = "Reads per barcode", main = paste0(libtype, " Reads per barcode"), col="darkblue", pch=16,ylim=c(1,10000))
+data <- data.frame(1:length(Size2), Size2)
+colnames(data) <- c("x","y")
+elbow <- elbow_knee_finder(data$x, log10(data$y))
+top_ranked <- factor(ifelse(data$x <= elbow[1], 1, 0))
+plot(Size2,
+     log="y", 
+     xlab=paste0(" Barcode rank (", length(Size2)-elbow[1], " low quality cells)"), 
+     ylab = "Reads per barcode", main = paste0(libtype, " Reads per barcode"), 
+     col=c("dimgrey","darkblue")[top_ranked], 
+     pch=16,
+     ylim=c(1,100000))
 abline(v=elbow[1], h=10^(elbow[2]))
 text(elbow[1], 10^(elbow[2]),
      paste("(", elbow[1], ", ", 10^(elbow[2]), ")", sep=""),
      adj=c(-0.1,-0.5))
 
-plot(Size2[1:elbow[1]], log="y", xlab="Barcode rank", ylab = "Reads per barcode", main = paste0(libtype, " Reads per (top ranked) barcode"), col="darkblue", pch=16, ylim=c(1,10000))
-elbow <- elbow_finder(1:elbow[1], log10(Size2[1:elbow[1]]))
-abline(v=elbow[1], h=10^(elbow[2]))
-text(elbow[1], 10^(elbow[2]),
-     paste("(", elbow[1], ", ", 10^(elbow[2]), ")", sep=""),
+# plot top ranked barcodes
+smooth_spline <- smooth.spline(x=data$x[1:elbow[1]], y=log10(data$y[1:elbow[1]]), spar=1)
+second_deriv <- predict(smooth_spline, x=data$x[1:elbow[1]], deriv=2)
+
+# if second derivative is all pos/all neg, plot is not "bumpy" so can use basic elbow/knee finder.
+# second deriv vals can be noisy at beginning and end of graph; exclude first 10% and last 10% 
+# of values when establishing uniformity of second deriv sign.
+# else, look for zeros of second deriv plot surrounding abs min of second deriv;
+# use these zeros as endpoints for elbow/knee finder function
+ten_percent <- round(length(second_deriv$x)*0.1)
+mid_second_deriv <- second_deriv$y[(ten_percent+1):(length(second_deriv$y)-ten_percent)] 
+if (all(mid_second_deriv >= 0) | all(mid_second_deriv <= 0)){
+  print("Finding elbow of top ranked barcodes")
+  point <- elbow_knee_finder(x_values=data$x[1:elbow[1]],
+                             y_values=log10(data$y[1:elbow[1]]))
+} else {
+  # find abs min
```

Comment:
This is a comment that human will read, please don't use abbreviations even if common.